src/Tasks/Microsoft.Common.CurrentVersion.targets

```diff
@@ -720,6 +720,120 @@ Copyright (C) Microsoft Corporation. All rights reserved.
+    
+    Include the final backslash in the path of each SourceRoot item.
+    
+    Extensibility: Add target to InitializeSourceRootDependsOn that populates SourceRoot item group.
+  -->
+
+  <Target Name="InitializeSourceRoot"
+          DependsOnTargets="$(InitializeSourceRootDependsOn)">
+    <ItemGroup>
+      <_InvalidSourceRoot Include="@(SourceRoot)" Condition="!HasTrailingSlash('%(SourceRoot.Identity)')" />
+    </ItemGroup>
+
+    <!--
+      The SourceRoot is required to have a trailing directory separator.
+      We do not append one implicitly as we do not know which separator to append on Windows.
+      The usage of SourceRoot might be sensitive to what kind of separator is used (e.g. in SourceLink where it needs
+      to match the corresponding separators used in paths given to the compiler).
+    -->
+    <Error Text="SourceRoot paths are required to end with a slash or backslash: @(_InvalidSourceRoot, ', ')"
+           Condition="'@(_InvalidSourceRoot)' != ''" />
+
+    <!--
+      Assign top-level source roots a deterministic path (MappedPath metadata).
+    -->
+
+    <ItemGroup Condition="'@(_MappedSourceRoot)' != ''">
+      <_MappedSourceRoot Remove="@(_MappedSourceRoot)" />
+    </ItemGroup>
+
+    <MapSourceRoots SourceRoots="@(SourceRoot)" Condition="'$(DeterministicSourcePaths)' == 'true'">
+      <Output TaskParameter="MappedSourceRoots" ItemName="_MappedSourceRoot" />
+    </MapSourceRoots>
+
+    <ItemGroup Condition="'$(DeterministicSourcePaths)' == 'true'">
+      <SourceRoot Remove="@(SourceRoot)" />
+      <SourceRoot Include="@(_MappedSourceRoot)" />
+    </ItemGroup>
+  </Target>
+
+  <!-- 
+    ================
+    SourceRevisionId
+    ================
+    
+    SourceRevisionId is embedded in AssemblyInformationalVersion, nuspec files for NuGet packages produced by the project, etc., 
+    to associate assemblies, packages and other build outputs with a particular revision of the repository from which they were built.
+   
+    The format of the id is specific to the source control system the project is built from. A commit hash for GIT repositories, 
+    changeset number for TFVC repositories, etc. 
+   
+    Here, the id is set to the value supplied by the CI server building the project, provided that the server follows 
+    Standard CI specification: https://github.com/dotnet/designs/blob/86f4ed0e39fc1b1ab5c4128990b17c4aead4420f/proposed/standard-ci-env-variables.md.
+      
+    If the project references a package that provides source control information (such as Microsoft.Build.Tasks.Git, Microsoft.Build.Tasks.TFVC, etc.)
+    the revision id becomes available in local build environments as well.
+    
+    Extensibility: Add target to InitializeSourceRevisionIdDependsOn to initialize SourceRevisionId.
+  -->
+
+  <Target Name="InitializeSourceRevisionId"
+          DependsOnTargets="_InitializeSourceRevisionIdFromStandardCI;$(InitializeSourceRevisionIdDependsOn)">
```

Comment:
Should `_InitializeSourceRevisionIdFromStandardCI` go after `$(InitializeSourceRevisionIdDependsOn)`?  That way other extensions can hook into this target, and they don't have to blindly overwrite the property?

As it is coded now, if an extension wants to write to `<SourceRevisionId>`, it can't check for empty, because `_InitializeSourceRevisionIdFromStandardCI` will set it first.

Same applies to `InitializePrivateRepositoryUrl` below.

Reply:
That depends on whether STANDARD_CI_SOURCE_REVISION_ID should be the default value that should be used if available or a fallback value that should be used if there is no extension setting SourceRevisionId, right?

Reply:
The extension could decide to chain their target before `InitializeSourceRevisionIdFromStandardCI` target to initialize to a different default value like so:

```
<Target Name="MyInitializeSourceRevisionIdFromStandardCI"
              BeforeTargets="InitializeSourceRevisionIdFromStandardCI">
    <PropertyGroup>
      <SourceRevisionId Condition="'$(SourceRevisionId)' == ''">MyRevisionId</SourceRevisionId>
    </PropertyGroup>
</Target>
```

Reply:
Think about if you were writing an extension to this target and you wanted to set `<SourceRevisionId>`.  How would you do it?

Maybe instead of using a `Target` to set this property, you should just set the property in a normal `<PropertyGroup>`.  Then the typical MSBuild ordering applies.  An extension can set it either before OR after you.  Here they only get 1 option - after.

Reply:
I have a working extension that does it: https://github.com/tmat/repository-info/blob/master/src/Microsoft.Build.Tasks.SourceControl/build/Microsoft.Build.Tasks.SourceControl.targets#L62-L67

It can't be set in normal `<PropertyGroup>` since a task has to be called to retrieve this value.

Reply:
Why have both extension mechanisms?  Why not just use `BeforeTargets` and `AfterTargets`?  And not have `InitializeSourceRevisionIdDependsOn` at all?

Reply:
`AfterTargets` doesn't work for the consumption. If I need to use this value in another target I add `InitializeSourceRevisionId` to my target's `DependsOnTargets`. `AfterTargets` of `InitializeSourceRevisionId` won't necessarily run.

Reply:
> Why have both extension mechanisms?

"... depends on whether STANDARD_CI_SOURCE_REVISION_ID should be the default value that should be used if available or a fallback value that should be used if there is no extension setting SourceRevisionId"

Reply:
I guess we could also have `InitializeSourceRevisionIdDependsOn` initialized to `_InitializeSourceRevisionIdFromStandardCI` and then let extensions to prepend or append their target name to this property.

Reply:
@rainersigwald Is there any guidelines for how to write these extension points?

Reply:
No, no formal guidelines that I'm aware of. I added some comments about the structure I'd expect, which I think is closer to what Eric expected too.

In general, I'd rather avoid creating new `xDependsOn` properties. They were required for MSBuild 2, but BeforeTargets/AfterTargets is a much better extensibility mechanism.

Reply:
OK. It would be nice to have these guidelines written down somewhere, so we don't need to spend time figuring this out every time.

Reply:
By the way, this

> `AfterTargets` doesn't work for the consumption. If I need to use this value in another target I add `InitializeSourceRevisionId` to my target's `DependsOnTargets`. `AfterTargets` of `InitializeSourceRevisionId` won't necessarily run.

Isn't true. It should be (targets should make a DAG), but it's not. The details of the traversal are [https://docs.microsoft.com/en-us/visualstudio/msbuild/target-build-order#determining-the-target-build-order](documented), and I don't see any way to change them to a DAG approach without breaking the whole world, so they can be relied upon.